### ISSUE TITLE
Crash with sorted scale in group mark (after cross+facet+signal change)

### ISSUE THREAD

**Issue by Jerrythafast** [vega/vega#566]
Hey guys,

I found a curious bug again. An MWE (trimmed down from over 400 lines, getting it this short was a hell of a job!):
```json
{
  "signals": [
    {"name": "filter_series", "init": 1}
  ],
  "data": [
    {"name": "x", "values": [1]},
    {
      "name": "a",
      "values": [
        {"series": 1, "groupname": "alpha"},
        {"series": 2, "groupname": "beta"},
        {"series": 4, "groupname": "beta"}
      ],
      "transform": [
        {
          "type": "filter",
          "test": "datum.series & filter_series"
        },
        {"type": "cross", "with": "x"},
        {"type": "facet", "groupby": ["a.groupname"]}
      ]
    }
  ],
  "marks": [
    {
      "type": "group",
      "from": {"data": "a"},
      "scales": [
        {
          "name": "c",
          "type": "ordinal",
          "range": "category20",
          "domain": {
            "field": "a.series",
            "sort": {"field": "a.series", "op": "min"}
          }
        }
      ]
    }
  ]
}
```
What we have here is three data points in two groups (named `alpha` and `beta`), and a signal to filter the data points. Initially, only the first data point passes the filter. The remaining data points are crossed with another dataset (in this MWE, a dummy dataset that only contains a `1`) and then faceted based on their `groupname`s. Finally, in the group mark, each facet gets a separate colour scale with a sorted domain.

Something goes wrong when the value of the `filter_series` signal is changed. Let's first change its value to `3` (causing the second data point to pass the filter as well):
```js
ved.view.signal("filter_series", 3).update()
```
Although it's impossible to tell from the outside (everything seems to work fine, ~~even with the complete 400-line spec I distilled this from~~*), things already get messy on the inside. I investigated this a little bit deeper and found that somewhere the `Collector.add()` method gets called with as its argument a tuple that was already added to that `Collector` before.

If we now change the value of the signal to `5` (causing the second data point to fail at the filter, but at the same time causing the third to pass):
```js
ved.view.signal("filter_series", 5).update()
```
Vega crashes because some variable (that was supposed to contain a data tuple) is `undefined`. I traced down the root cause of this, which turned out to be that the [`Collector.values()` method](https://github.com/vega/datalib/blob/master/src/aggregate/collector.js#L23) defines the length of its output array as `this._add.length - this._rem.length`. And because some (or all) data tuples were `add`ed twice, it ends up overestimating the length of its output array (which gets padded to that length with `undefined`). Some other code that is looping over the return value of `Collector.values()` will then crash once it tries to access a property of `undefined`.

I highly doubt Vega was supposed to call `add` with the same argument twice, so I guess ultimately that's what needs to be fixed.

Also,
- I was unable to reproduce this without the `cross`
- You don't need to sort the scale, but if you don't, Vega crashes only after flipping the signal value between `3` and `5` twice

Just like with vega/vega-scenegraph#24, closer analysis revealed that the crash occurs only if `fullUpdate` was `true` in [GroupBuilder.js lines 76-77](https://github.com/vega/vega/blob/master/src/scene/GroupBuilder.js#L76), which was introduced in @e2afbe3. Although this crash seems very different from that one, that might hint at a common cause (but it might also be a dead lead, you know Vega better than I do :wink: ).

Hope this gives enough information for you to track down the bug!

Kind regards,
Jerry

\***EDIT**: Now that I come to think of it, actually some misbehaviour _is_ visible in the full spec. The colour scale is updated incorrectly: it is not removing the values that are filtered out after the signal change. Reverting back to version `2.5.0` shows the same thing, but without crashing. In `2.5.0`, values are removed from the scale once I change the signal another time to yet a different value. In short, the scale always contains those values that match either the current filter _or the previous filter_.

**Comment by arvind** [comment]
Hi @Jerrythafast, fantastic bug report again! I think I may have caught a similar bug a couple of weeks ago. Do you mind manually applying the fix from 16e521f and seeing if it resolves your issue? I'm no longer able to get Vega to crash with your MWE above.

**Comment by Jerrythafast** [comment]
Hey @arvind, thanks for the response. That indeed seems to fix it. I forgot to check whether there were newer commits after the release of `2.5.2`. Could have saved me a lot of time and effort :disappointed_relieved: 

The commit you linked also appears to fix the root cause of the crash in vega/vega-scenegraph#24. I'm not sure whether that patch is still appropriate, though. I'll leave it up to the Vega team to decide whether to merge that patch or to reject it.

**Comment by Jerrythafast** [comment]
The strange behaviour with the colour scale that I was having (and still have) in the complete spec seems to be the same issue as #446. I tried to make a MWE for this and while doing that the horizontal axis disappeared again. Even better still, I managed to get Vega to put the Legend items (marks and texts) in the same SVG `<g>` element as the ticks and texts of the horizontal axis!

**Comment by Jerrythafast** [comment]
This is now confirmed fixed in version 2.6.0. (Though the same spec is still giving me #446 oddities)
